crates/sovereign-sdk/module-system/sov-state/src/witness.rs

```diff
@@ -17,28 +20,54 @@ use sov_modules_core::Witness;
 /// assert_eq!(witness.get_hint::<u64>(), 1u64);
 /// assert_eq!(witness.get_hint::<u64>(), 2u64);
 /// ```
-#[derive(Default, BorshDeserialize, BorshSerialize, Debug, Serialize, Deserialize)]
+#[derive(Default, Debug, Serialize, Deserialize)]
 pub struct ArrayWitness {
-    next_idx: usize,
-    hints: Vec<Vec<u8>>,
+    next_idx: AtomicUsize,
+    hints: Mutex<Vec<Vec<u8>>>,
 }
 
 impl Witness for ArrayWitness {
-    fn add_hint<T: BorshSerialize>(&mut self, hint: T) {
-        self.hints.push(borsh::to_vec(&hint).unwrap())
+    fn add_hint<T: BorshSerialize>(&self, hint: T) {
+        self.hints
+            .lock()
+            .unwrap()
+            .push(borsh::to_vec(&hint).unwrap());
     }
 
-    fn get_hint<T: BorshDeserialize>(&mut self) -> T {
-        let idx = self.next_idx;
-        self.next_idx += 1;
-        T::deserialize_reader(&mut std::io::Cursor::new(&self.hints[idx]))
+    fn get_hint<T: BorshDeserialize>(&self) -> T {
+        let idx = self
+            .next_idx
+            .fetch_add(1, std::sync::atomic::Ordering::SeqCst);
+        let hints_lock = self.hints.lock().unwrap();
+        T::deserialize(&mut hints_lock[idx].as_slice())
             .expect("Hint deserialization should never fail")
     }
 
-    fn merge(&mut self, rhs: &mut Self) {
-        let rhs_next_idx = rhs.next_idx;
-        let lhs_hints_lock = &mut self.hints;
-        let rhs_hints_lock = &mut rhs.hints;
-        lhs_hints_lock.extend(rhs_hints_lock.drain(rhs_next_idx..))
+    fn merge(&self, rhs: &Self) {
+        let rhs_next_idx = rhs.next_idx.load(std::sync::atomic::Ordering::SeqCst);
+        let mut lhs_hints_lock = self.hints.lock().unwrap();
+        let mut rhs_hints_lock = rhs.hints.lock().unwrap();
+        lhs_hints_lock.extend(rhs_hints_lock.drain(rhs_next_idx..));
+    }
+}
+
+impl BorshSerialize for ArrayWitness {
+    fn serialize<W: std::io::prelude::Write>(&self, writer: &mut W) -> std::io::Result<()> {
+        let hints = self.hints.lock().unwrap();
+        borsh::BorshSerialize::serialize(
+            &self.next_idx.load(std::sync::atomic::Ordering::Relaxed),
+            writer,
+        )?;
+        borsh::BorshSerialize::serialize(&*hints, writer)?;
+        Ok(())
```

Comment:
Lock the hints after serializing next_idx, it will be more efficient as you won't be blocking hints unnecessarily while serializing next_idx.